crates/nu-protocol/src/io/lines.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{IntoSpanned, ShellError, Span, Value};
+use std::{
+    io::{self, BufRead, BufReader, Read},
+    sync::{atomic::AtomicBool, Arc},
+};
+
+pub struct ByteLines<R: Read>(BufReader<R>);
+
+impl<R: Read> ByteLines<R> {
+    pub fn new(read: R) -> Self {
+        Self(BufReader::new(read))
+    }
+}
+
+impl<R: Read> Iterator for ByteLines<R> {
+    type Item = io::Result<Vec<u8>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut buf = Vec::new();
+        // `read_until` will never stop reading unless `\n` or EOF is encountered,
+        // so we may want to limit the number of bytes by using `take` as the Rust docs suggest.
+        // let capacity = self.0.capacity() as u64;
+        // let mut reader = (&mut self.0).take(capacity);
+        let reader = &mut self.0;
+        match reader.read_until(b'\n', &mut buf) {
+            Ok(0) => None,
+            Ok(_) => {
+                if buf.ends_with(&[b'\n']) {
+                    let _ = buf.pop();
+                    if buf.ends_with(&[b'\r']) {
+                        let _ = buf.pop();
+                    }
+                }
+                Some(Ok(buf))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+pub struct Lines<R: Read> {
+    lines: ByteLines<R>,
+    span: Span,
+    ctrlc: Option<Arc<AtomicBool>>,
+}
+
+impl<R: Read> Lines<R> {
+    pub fn new(reader: R, span: Span, ctrlc: Option<Arc<AtomicBool>>) -> Self {
+        Self {
+            lines: ByteLines::new(reader),
+            span,
+            ctrlc,
+        }
+    }
+
+    pub fn span(&self) -> Span {
+        self.span
+    }
+}
+
+impl<R: Read> Iterator for Lines<R> {
+    type Item = Result<Vec<u8>, ShellError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if nu_utils::ctrl_c::was_pressed(&self.ctrlc) {
+            None
+        } else {
+            match self.lines.next() {
+                Some(Ok(line)) => Some(Ok(line)),
+                Some(Err(err)) => Some(Err(err.into_spanned(self.span).into())),
+                None => None,
+            }
+        }
+    }
+}
+
+pub struct Values<R: Read> {
+    lines: ByteLines<R>,
```

Comment:
This is something I've already changed on my branch, but just so you know, I don't think `Values` should really be lines. I think it's currently just whatever the buffers coming in are, and it's additionally split up by lines to kind of do a pseudo-line-buffering, but that's a source of bugs.

Changing it to always be the full line is okay for text data, but not ideal for binary data because the chunks are all going to be random sizes / potentially the whole file because you never really know when you'll get a `\n` character.

If you want to check out my branch instead, I made it basically use `BufRead::fill_buf` to get whatever data is pending and try to create valid UTF-8 buffers when in string mode. It's quite obviously different from the output of `lines`, so that should help to force anyone who actually wants lines to use that.

Reply:
Hmm, yeah the `fill_buf` approach does sound better. Ideally, `values` should not be needed / should not exist on `ByteStream`, and commands would just use the reader directly (or maybe `lines`). Unfortunately, we have a bunch of commands that don't do that and/or use one of the `PipelineData` methods (`map`, `filter`, `flat_map`, `into_iter_strict`, the `IntoIterator` trait, etc.).